File: Chapter_3/06_solution.py
import os
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry


class TwilioAPI:
    def __init__(self, account_sid=None, auth_token=None, base_url="https://api.twilio.com/2010-04-01"):
        """
        Initialize the TwilioAPI class.

        :param account_sid: Twilio Account SID
        :param auth_token: Twilio Auth Token
        :param base_url: Base URL for the Twilio API
        """
        self.account_sid = account_sid or os.getenv("TWILIO_ACCOUNT_SID")
        self.auth_token = auth_token or os.getenv("TWILIO_AUTH_TOKEN")
        self.base_url = f"{base_url}/Accounts/{self.account_sid}"
        self.headers = {
            "Authorization": f"Basic {self._encode_credentials()}",
            "Content-Type": "application/x-www-form-urlencoded"
        }
        self.session = requests.Session()
        retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])
        self.session.mount("https://", HTTPAdapter(max_retries=retries))

    def _encode_credentials(self):
        """
        Encode the account SID and auth token for basic authentication.

        :return: Encoded credentials
        """
        credentials = f"{self.account_sid}:{self.auth_token}"
        return base64.b64encode(credentials.encode()).decode()

    def _request(self, method, url, data=None):
        """
        Make an HTTP request.

        :param method: HTTP method (GET, POST, etc.)
        :param url: URL for the request
        :param data: Data to send with the request
        :return: JSON response
        """
        response = self.session.request(method, url, headers=self.headers, data=data)
        response.raise_for_status()
        return response.json()


class TwilioMessagingAPI(TwilioAPI):
    def send_message(self, to, from_, body):
        """
        Send an SMS message.

        :param to: The destination phone number
        :param from_: The phone number to send the message from
        :param body: The body of the message
        :return: The response from the API
        """
        url = f"{self.base_url}/Messages.json"
        data = {
            'To': to,
            'From': from_,
            'Body': body
        }
        return self._request("POST", url, data)

    def fetch_message(self, message_sid):
        """
        Fetch a message.

        :param message_sid: The SID of the message to fetch
        :return: The response from the API
        """
        url = f"{self.base_url}/Messages/{message_sid}.json"
        return self._request("GET", url)

    def list_messages(self, params=None):
        """
        List messages.

        :param params: Optional parameters for filtering the list
        :return: The response from the API
        """
        url = f"{self.base_url}/Messages.json"
        if params:
            url += '?' + '&'.join([f"{key}={value}" for key, value in params.items()])
        return self._request("GET", url)


# Example usage:
if __name__ == "__main__":
    account_sid = "your_account_sid"
    auth_token = "your_auth_token"
    twilio = TwilioMessagingAPI(account_sid, auth_token)

    try:
        # Send an SMS message
        response = twilio.send_message("+1234567890", "+0987654321", "Hello, World!")
        print(response)

        # Fetch a message
        message = twilio.fetch_message("SMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")
        print(message)

        # List messages
        messages = twilio.list_messages({'PageSize': 20})
        print(messages)
    except requests.exceptions.HTTPError as e:
        print(f"HTTP error occurred: {e}")
    except Exception as e:
        print(f"An error occurred: {e}")

File: Chapter_3/06_solution.php
<?php

/**
 * Twilio Messaging API Client
 *
 * This class provides methods to interact with the Twilio Messaging API.
 *
 * @category  API
 * @package   TwilioMessagingAPI
 * @license   https://opensource.org/licenses/MIT MIT License
 * @link      https://www.twilio.com/docs/sms
 */

class TwilioMessagingAPI
{
    private $accountSid;
    private $authToken;
    private $baseUrl;
    private $headers;

    /**
     * Constructor
     *
     * @param string $accountSid Account SID
     * @param string $authToken  Auth Token
     */
    public function __construct($accountSid, $authToken)
    {
        $this->accountSid = $accountSid;
        $this->authToken = $authToken;
        $this->baseUrl = "https://api.twilio.com/2010-04-01/Accounts/{$this->accountSid}";
        $this->headers = [
            "Authorization: Basic " . base64_encode("{$this->accountSid}:{$this->authToken}"),
            "Content-Type: application/x-www-form-urlencoded"
        ];
    }

    /**
     * Make an HTTP request
     *
     * @param string $method HTTP method (GET, POST, etc.)
     * @param string $url    URL for the request
     * @param array  $data   Data to send with the request
     *
     * @return array JSON response
     * @throws Exception If an HTTP error occurs
     */
    private function request($method, $url, $data = null)
    {
        $ch = curl_init();

        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $this->headers);

        if ($data) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($data));
        }

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

        if ($httpCode >= 400) {
            throw new Exception("HTTP Error: $httpCode - Response: $response");
        }

        curl_close($ch);

        return json_decode($response, true);
    }

    /**
     * Send an SMS message
     *
     * @param string $to   The destination phone number
     * @param string $from The phone number to send the message from
     * @param string $body The body of the message
     *
     * @return array The response from the API
     * @throws Exception If an HTTP error occurs
     */
    public function sendMessage($to, $from, $body)
    {
        $url = "{$this->baseUrl}/Messages.json";
        $data = [
            'To' => $to,
            'From' => $from,
            'Body' => $body
        ];

        return $this->request("POST", $url, $data);
    }

    /**
     * Fetch a message
     *
     * @param string $messageSid The SID of the message to fetch
     *
     * @return array The response from the API
     * @throws Exception If an HTTP error occurs
     */
    public function fetchMessage($messageSid)
    {
        $url = "{$this->baseUrl}/Messages/{$messageSid}.json";

        return $this->request("GET", $url);
    }

    /**
     * List messages
     *
     * @param array $params Optional parameters for filtering the list
     *
     * @return array The response from the API
     * @throws Exception If an HTTP error occurs
     */
    public function listMessages($params = [])
    {
        $url = "{$this->baseUrl}/Messages.json";

        if (!empty($params)) {
            $url .= '?' . http_build_query($params);
        }

        return $this->request("GET", $url);
    }
}

// Example usage:
$accountSid = "your_account_sid";
$authToken = "your_auth_token";
$twilio = new TwilioMessagingAPI($accountSid, $authToken);

try {
    // Send an SMS message
    $response = $twilio->sendMessage("+1234567890", "+0987654321", "Hello, World!");
    print_r($response);

    // Fetch a message
    $message = $twilio->fetchMessage("SMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX");
    print_r($message);

    // List messages
    $messages = $twilio->listMessages(['PageSize' => 20]);
    print_r($messages);
} catch (Exception $e) {
    echo 'Error: ' . $e->getMessage();
}